pkg/bootconfig/bootconfig.go

```diff
@@ -3,69 +3,97 @@ package bootconfig
 import (
 	"encoding/json"
 	"errors"
+	"fmt"
 	"log"
 	"os"
 
 	"github.com/systemboot/systemboot/pkg/crypto"
 	"github.com/u-root/u-root/pkg/kexec"
+	"github.com/u-root/u-root/pkg/multiboot"
 )
 
 // BootConfig is a general-purpose boot configuration. It draws some
 // characteristics from FIT but it's not compatible with it. It uses
 // JSON for interoperability.
 type BootConfig struct {
-	Name       string `json:"name,omitempty"`
-	Kernel     string `json:"kernel"`
-	Initramfs  string `json:"initramfs,omitempty"`
-	KernelArgs string `json:"kernel_args,omitempty"`
-	DeviceTree string `json:"devicetree,omitempty"`
+	Name          string   `json:"name,omitempty"`
+	Kernel        string   `json:"kernel"`
+	Initramfs     string   `json:"initramfs,omitempty"`
+	KernelArgs    string   `json:"kernel_args,omitempty"`
+	DeviceTree    string   `json:"devicetree,omitempty"`
+	Multiboot     string   `json:"multiboot_kernel,omitempty"`
+	MultibootArgs string   `json:"multiboot_args,omitempty"`
+	Modules       []string `json:"multiboot_modules,omitempty"`
 }
 
 // IsValid returns true if a BootConfig object has valid content, and false
 // otherwise
 func (bc *BootConfig) IsValid() bool {
-	return bc.Kernel != ""
+	return (bc.Kernel != "" && bc.Multiboot == "") || (bc.Kernel == "" && bc.Multiboot != "")
 }
 
 // Boot tries to boot the kernel with optional initramfs and command line
 // options. If a device-tree is specified, that will be used too
 func (bc *BootConfig) Boot() error {
-	crypto.TryMeasureBootConfig(bc.Name, bc.Kernel, bc.Initramfs, bc.KernelArgs, bc.DeviceTree)
-
-	kernel, err := os.Open(bc.Kernel)
-	if err != nil {
-		return err
-	}
-	var initramfs *os.File
-	if bc.Initramfs != "" {
-		initramfs, err = os.Open(bc.Initramfs)
+	data := bc.Name + bc.Kernel + bc.Initramfs + bc.KernelArgs + bc.DeviceTree + bc.Multiboot + bc.MultibootArgs
+	crypto.TryMeasureData(crypto.BootConfigPCR, []byte(data), "bootconfig")
```

Comment:
Doesn't this produce a different result than previously with `TryMeasureBootConfig` calling `Measure` separately on each string?

Reply:
Now it is measured in the same way as it was intended by MeasureBootConfig()